scipy/sparse/linalg/isolve/tests/test_lsmr.py

```diff
@@ -97,15 +97,30 @@ def testColumnB(self):
 
     def testInitialization(self):
         # Test that the default setting is not modified
-        x_ref = lsmr(G, b)[0]
+        x_ref, _, itn_ref, normr_ref, *_ = lsmr(G, b)
+        assert_allclose(norm(b - G@x_ref), normr_ref, atol=1e-6);
+
+        # Test passing zeros yields similiar result
         x0 = zeros(b.shape)
         x = lsmr(G, b, x0=x0)[0]
-        assert_array_almost_equal(x_ref, x)
+        assert_array_almost_equal(x, x_ref)
 
         # Test warm-start with single iteration
         x0 = lsmr(G, b, maxiter=1)[0]
-        x = lsmr(G, b, x0=x0)[0]
-        assert_array_almost_equal(x_ref, x)
+
+        # NOTE(gh-12139): This doesn't always converge to the same value as
+        # ref because error estimates will be slightly different when calculated
+        # from zeros vs x0 as a result only compare norm and itn (not x).
+        x, _, itn, normr, *_ = lsmr(G, b, x0=x0)
+        assert_allclose(norm(b - G@x), normr, atol=1e-6);
+
+        # One less or same number of iterations.
+        assert itn_ref - itn in (-1, 0)
+
+        # If an extra iteration is performed normr may be 0, while normr_ref
+        # may be much larger.
+        assert_allclose(normr, normr_ref, atol=1e-6, rtol=1);
```

Comment:
this might be better expressed as `assert normr < normr_ref * (1 + 1e-6)`

`normr` had equal or greater iterations so 99.999% of the time it has equal or smaller error.

Reply:
agreed